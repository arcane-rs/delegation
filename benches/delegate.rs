use criterion::{criterion_group, criterion_main, Criterion};
use delegation::delegate;

#[delegate]
trait AsStr {
    fn as_str(&self) -> &str;
}

impl AsStr for String {
    fn as_str(&self) -> &str {
        self
    }
}

macro_rules! def_enum {
    ($($variant:ident),+) => {
        #[delegate(derive(AsStr))]
        enum Name {
            $($variant(String)),+
        }
    };
}

macro_rules! def_bench {
    ($($variant:ident),*) => {
        fn bench(bencher: &mut ::criterion::Bencher) {
            $(
                bencher.iter(|| {
                    let name = ::criterion::black_box(
                        Name::$variant(String::from("Test"))
                    );

                    assert_eq!(name.as_str(), "Test");
                });
            )*
        }
    };
}

macro_rules! define {
    ($($variant:ident),*) => {
        def_enum!($($variant),*);
        def_bench!($($variant),*);
    };
}

define!(
    A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y,
    Z
);

fn delegate_benchmark(c: &mut Criterion) {
    c.bench_function("delegate", bench);
}

criterion_group!(benches, delegate_benchmark);
criterion_main!(benches);
