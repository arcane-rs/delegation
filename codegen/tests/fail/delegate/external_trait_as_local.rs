use delegation_codegen::delegate;

#[delegate(as = AsRef)]
pub trait AsRefDef<T: ?Sized> {
    fn as_ref(&self) -> &T;
}

#[delegate(derive(AsRefDef<str>))]
pub enum Name {
    First(String),
}

fn main() {
    unreachable!()
}
