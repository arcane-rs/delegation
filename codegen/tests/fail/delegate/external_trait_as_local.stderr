error[E0407]: method `as_ref` is not a member of trait `AsRefDef`
 --> tests/errors/external_trait_as_local.rs:5:5
  |
5 |     fn as_ref(&self) -> &T;
  |     ^^ not a member of trait `AsRefDef`
...
8 | #[delegate(derive(for<> AsRefDef<str>))]
  | ---------------------------------------- in this procedural macro expansion
  |
  = note: this error originates in the macro `AsRefDef` which comes from the expansion of the attribute macro `delegate` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0576]: cannot find method or associated constant `as_ref` in trait `AsRefDef`
 --> tests/errors/external_trait_as_local.rs:5:8
  |
5 |     fn as_ref(&self) -> &T;
  |        ^^^^^^ not found in `AsRefDef`
...
8 | #[delegate(derive(for<> AsRefDef<str>))]
  | ---------------------------------------- in this procedural macro expansion
  |
  = note: this error originates in the macro `AsRefDef` which comes from the expansion of the attribute macro `delegate` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0046]: not all trait items implemented, missing: `Wrapper`, `__delegate_AsRefDef__Bind0`
 --> tests/errors/external_trait_as_local.rs:3:1
  |
3 | #[delegate(as = AsRef)]
  | ^^^^^^^^^^^^^^^^^^^^^^^
  | |
  | missing `Wrapper`, `__delegate_AsRefDef__Bind0` in implementation
  | `Wrapper` from trait
  | `__delegate_AsRefDef__Bind0` from trait
...
8 | #[delegate(derive(for<> AsRefDef<str>))]
  | ---------------------------------------- in this procedural macro expansion
  |
  = note: this error originates in the macro `AsRefDef` which comes from the expansion of the attribute macro `delegate` (in Nightly builds, run with -Z macro-backtrace for more info)
